app/main/views/suppliers.py

```diff
@@ -88,6 +90,112 @@ def list_suppliers():
         abort(400, 'invalid framework')
 
 
+@main.route('/suppliers/export/<framework_slug>', methods=['GET'])
+def export_suppliers_for_framework(framework_slug):
+    # 400 if framework slug is invalid
+    framework = Framework.query.filter(Framework.slug == framework_slug).first()
+    if not framework:
+        abort(400, 'invalid framework')
+
+    if framework.status == 'coming':
+        abort(400, 'framework not yet open')
+
+    lot_slugs_by_id = {lot.id: lot.slug for lot in framework.lots}
+
+    # Creates a dictionary of dictionaries, with published service counts per lot per supplier.
+    # To do this we use dictionary comprehension, SQL query and itertools' 'groupby' function.
+    # Grouping by supplier id is more efficient as this is what we will be iterating over later.
+    service_counts_by_lot_by_supplier = {
+        supplier_id: {row[1]: row[2] for row in rows}
+        for supplier_id, rows in groupby(
```

Comment:
The variable rows here isn't very descriptive. It feels like we're unessecarily complicating things by trying to do this in one line.

If we assign the query to a separate variable we can be more descriptive about what it is:

```
    supplier_id_lot_id_service_count_tuples = db.session.query(
        Service.supplier_id,
        Service.lot_id,
        func.count(Service.id)
    ).filter(
        Service.status == 'published',
        Service.framework_id == framework.id
    ).group_by(
        Service.supplier_id,
        Service.lot_id,
    ).all()
```

Reply:
Then the comprehension/ groupby also seems a bit over involved and has a lot of `get`s used for setting default values. 
There is a method in collections called defaultdict that you can use to set defaults when unpacking stuff https://docs.python.org/2/library/collections.html#collections.defaultdict

The usage here would look something like this:
```
    # Convert the above to a dictionary like {supplier_id: {lot_1_slug: lot_1_count, lot_2_slug: lot_2_count}}
    service_counts_by_lot_by_supplier_id = defaultdict(dict)
    for supplier_id, lot_id, service_count in supplier_id_lot_id_service_count_tuples:
        service_counts_by_lot_by_supplier_id.setdefault(supplier_id, {}).update({lot_id: service_count})
```

Reply:
You could alos unpack the lot slug here to save doing it further down:
```
    # Convert the above to a dictionary like {supplier_id: {lot_1_slug: lot_1_count, lot_2_slug: lot_2_count}}
    service_counts_by_lot_by_supplier_id = defaultdict(dict)
    for supplier_id, lot_id, service_count in supplier_id_lot_id_service_count_tuples:
        service_counts_by_lot_by_supplier_id.setdefault(supplier_id, {}).update({lot_slugs_by_id[lot_id]: service_count})

```

Reply:
Then instead of the comprehension on line 181 you could set the defaults on that and away you go 🤷‍♀️

Reply:
I strongly disagree with all these assertions about style. This is how I used to write things many years ago and there's a reason I migrated away from it - it is far harder to understand what high level operation the user is trying to achieve. Written in a functional style, the code structure tends to mirror the structure of the resulting output There is no such link having to follow lines and lines of imperative code which could be doing anything. *Describe the cake to me*, don't read me the long recipe.

However I'm not going to continue this discussion as it's not fruitful and is just bikeshedding in the extreme. But I think it should be kept in mind that these suggestions are just your *opinion*, which is not necessarily agreed with.

Reply:
(or, if you want another argument and you still have faith in them, look at the mccabe scores of the two approaches...)